### PR TITLE
Dashboard bc

### PR DESCRIPTION
**Pull request by victor-paumier**
- add stats of section on dashboard for BC 
- add global stats on dashboard for admins
- add weekStats list
- add stats command for cron
- allow multiple select on lists

[Screen](https://github.com/ESNFranceG33kTeam/sf_buddysystem/files/3107377/screencapture-buddysystem-local-en-dashboard-2019-04-23-13_58_46.pdf)

### CODE REVIEW

File: assets/js/dashboard_insight.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        const dataMotivations = [];
+        $('.chart-motivations .chart-legend li').each(function () {
+            dataMotivations.push($(this).data('value'));
+        });
+        Chart.donutChart('.ct-chart-motivations', { series: dataMotivations });
+
+        if ($(universitiesSelector).length) {
+            $(document).on('change', universitiesSelector, () => {
+                const $legends = $('.chart-universities .chart-legend li');
+                const loop = parseInt($(universitiesSelector).val(), 10);
```
Comment (JeremieSamson) [inline assets/js/dashboard_insight.js]: not explicit enough, `const loopIndex` maybe ?

File: assets/js/dashboard_insight.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    _initCharts();
+
+    if ($statsSectionSelector.length) {
+        $statsSectionSelector.on('change', function () {
+            loader.start();
+
+            $.ajax({
+                type: 'GET',
```
Comment (JeremieSamson) [inline assets/js/dashboard_insight.js]: default ?

File: src/Entity/WeekStat.php [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+ * @ORM\Entity(repositoryClass="App\Repository\WeekStatRepository")
+ */
+class WeekStat
+{
+    use TimestampableEntity;
+
+    /**
+     * @var int
+     *
+     * @ORM\Column(type="integer")
```
Comment (JeremieSamson) [inline src/Entity/WeekStat.php]:
```suggestion
     * @ORM\Column(type="integer", options={"unsigned": true})
```

File: src/Entity/WeekStat.php [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+     * @ORM\Id
+     * @ORM\GeneratedValue
+     */
+    private $id;
+
+    /**
+     * @ORM\ManyToOne(targetEntity="App\Entity\Section", inversedBy="weekStats")
+     * @ORM\JoinColumn(name="section_id", referencedColumnName="id")
```
Comment (JeremieSamson) [inline src/Entity/WeekStat.php]: I'm not sure that this line is mandatory :thinking:

File: src/Repository/UserRepository.php [hunk 5]
```diff
@@ -54,7 +54,7 @@ public function getListStudentsNotMatched(
         Section $section = null,
         User $menteeToExclude = null
     ): array {
-        $qb = $this->getStudentsNotMatched($dataTableColumns->getSearch(), $section, $menteeToExclude);
+        $qb = $this->getStudentsToMatched($dataTableColumns->getSearch(), $section, $menteeToExclude);
```
Comment (JeremieSamson) [inline src/Repository/UserRepository.php]: typo, should be `getStudentsToMatch` IMO

File: src/Repository/UserRepository.php [hunk 6]
```diff
@@ -74,7 +74,7 @@ public function getListStudentsNotMatched(
 
     public function getNbUsersNotMatched(array $search = [], Section $section = null, User $menteeToExclude = null): int
     {
-        $qb = $this->getStudentsNotMatched($search, $section, $menteeToExclude);
+        $qb = $this->getStudentsToMatched($search, $section, $menteeToExclude);
```
Comment (JeremieSamson) [inline src/Repository/UserRepository.php]: same here

File: src/Repository/UserRepository.php [hunk 7]
```diff
@@ -297,29 +330,49 @@ private function getFilteredUsers(array $search = [], Section $section = null):
...
+                    ;
+                }
             }
         }
 
         return $qb;
     }
 
-    private function getStudentsNotMatched(
+    private function getStudentsToMatched(
```
Comment (JeremieSamson) [inline src/Repository/UserRepository.php]: can you rename this function please ?

File: src/Repository/WeekStatRepository.php [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+        $datesAvailable = array_map(function ($a) {return array_pop($a); }, $qbDates->getQuery()->getArrayResult());
+        if (\count($datesAvailable) > $maxResults) {
+            $dates = [$datesAvailable[0]];
+            unset($datesAvailable[0]);
+            $lastDate = $datesAvailable[\count($datesAvailable)];
+            unset($datesAvailable[\count($datesAvailable)]);
+            $datesAvailable = array_values($datesAvailable);
+
+            $maxResultsDate = $maxResults - 2;
+            $divider = (\count($datesAvailable) / $maxResultsDate);
```
Comment (JeremieSamson) [inline src/Repository/WeekStatRepository.php]:
```suggestion
            $divider = \count($datesAvailable) / $maxResultsDate;
```

File: src/Services/StatsService.php [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+        $this->localeManager = $localeManager;
+    }
+
+    public function generateNewStats(): void
+    {
+        $sections = $this->sectionRepo->findAll();
+
+        foreach ($sections as $section) {
+            $stat = new WeekStat();
+            $stat->setSection($section);
```
Comment (JeremieSamson) [inline src/Services/StatsService.php]:
```suggestion
            $section->addWeekStat($stat);
```

Comment (victor-paumier) [reply src/Services/StatsService.php]: done

File: src/Repository/WeekStatRepository.php (hunk 8 above)
Comment (victor-paumier) [reply src/Repository/WeekStatRepository.php]: done

File: src/Repository/UserRepository.php (hunk 7 above)
Comment (victor-paumier) [reply src/Repository/UserRepository.php]: done

File: src/Repository/UserRepository.php (hunk 6 above)
Comment (victor-paumier) [reply src/Repository/UserRepository.php]: done

File: src/Repository/UserRepository.php (hunk 5 above)
Comment (victor-paumier) [reply src/Repository/UserRepository.php]: done

File: src/Entity/WeekStat.php (hunk 4 above)
Comment (victor-paumier) [reply src/Entity/WeekStat.php]: done

File: src/Entity/WeekStat.php (hunk 3 above)
Comment (victor-paumier) [reply src/Entity/WeekStat.php]: done

File: assets/js/dashboard_insight.js (hunk 2 above)
Comment (victor-paumier) [reply assets/js/dashboard_insight.js]: done

File: assets/js/dashboard_insight.js (hunk 1 above)
Comment (victor-paumier) [reply assets/js/dashboard_insight.js]: done

File: assets/js/dashboard_insight.js [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+import Routing from './tools/Routing';
+import Flashbag from './tools/Flashbag';
+import loader from './tools/Loader';
+import '../sass/pages/dashboard_insight.scss';
+
+global.Chart = Chart;
+
+$(document).ready(() => {
+    const universitiesSelector = '#chart-universities__selector';
+    const $statsSectionSelector = $('#section-selector');
```
[the remark this answers is not in the document]
Comment (victor-paumier) [reply assets/js/dashboard_insight.js]: done

**Pull request merged**